File: topcoder/2402/BadNeighbors.py
class BadNeighbors:
    def __init__(self):
        pass

    def maxDonations(self, a):
        self.a = a
        self.m = [[0 for _ in enumerate(a)],[0 for _ in enumerate(a)]]
        return self.n(0,len(a)-1)

    def n(self, j,i):
        if i < j: 
            return 0
        elif self.m[j][i]: return self.m[j][i]
        elif i == j:
            result = self.a[i]
            self.m[j][i] = result
            return result
        elif i == j+1:
            result = max(self.a[i], self.a[j])
            self.m[j][i] = result
            return result
        elif i == len(self.a)-1:
            result = max(self.a[i]+self.n(1,i-2), self.n(0,i-1))
            self.m[j][i] = result
            return result
        else:
            result = max(self.a[i]+self.n(j,i-2), self.n(j, i-1))
            self.m[j][i] = result
            return result

File: topcoder/2402/BadNeighbors.cpp
/**
    @author: c4ts0up
    @language: C++
    @task: 2022-10 - MUA 7B
*/

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef long double ld;
#define pb push_back
#define ff first
#define ss second
#define endl "\n"

void fast_io() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
}

const ll MOD = 1e9+7, NMAX = 45, INF = 1e10;

class BadNeighbors {
    public:
        int n;
        vector <int> arr;
        int dp[NMAX][2][2];
        bool vis[NMAX][2][2];

        // empty constructor
        BadNeighbors() {}

        // principal
        int maxDonations(vector <int> sequence) {
            n = sequence.size();
            arr.resize(n);

            for (int i=0; i<n; i++) arr[i] = sequence[i];
            memset(vis, false, sizeof(vis));

            return play(0, false, false);
        }

        int play(int idx, bool prev, bool first_donated) {
            // caso base
            if (idx >= n) return 0LL;
            // caso especial último
            else if (idx == n-1 && first_donated) return 0LL;
            // caso dp
            else if (vis[idx][prev][first_donated]) return dp[idx][prev][first_donated];
            // caso recursivo
            else {
                ll take {-INF}, notake {-INF};

                // puedo tomarlo
                if (!prev) {
                    if (idx == 0) take = arr[idx] + play(idx+1, true, true);
                    else take = arr[idx] + play(idx+1, true, first_donated);
                }

                notake = play(idx+1, false, first_donated);

                vis[idx][prev][first_donated] = true;
                return dp[idx][prev][first_donated] = max(take, notake);
            }
        }
};